MinGW compiler was excluded for file related functions

**Describe the bug**
With commit https://github.com/vsg-dev/VulkanSceneGraph/commit/f1557223d050f386d2ed473c0300184d3f886250 several places were changed so that they are only activated when using a msvc compiler. These are the definition of the path separator and Windows-specific wide character and security-related functions. In the first case this leads to runtime errors, because paths are no longer separated correctly.  In the other cases, less secure and restricted functions are used.

**To Reproduce**
1. Inspect https://github.com/vsg-dev/VulkanSceneGraph/blob/affebd1aae1ccf71ab50a4d583b7e3ed2fe35f57/src/vsg/io/FileSystem.cpp#L55, which is only enabled when using msvc compiler.
2. Inspect https://github.com/vsg-dev/VulkanSceneGraph/commit/f1557223d050f386d2ed473c0300184d3f886250#diff-fde3bc681bc65905ab9792631af9ac2886ae7ca65228e4f83677cc0f508e1d3fR77, where the getenv_s function is not used with MinGW
3. Inspect https://github.com/vsg-dev/VulkanSceneGraph/commit/f1557223d050f386d2ed473c0300184d3f886250#diff-fde3bc681bc65905ab9792631af9ac2886ae7ca65228e4f83677cc0f508e1d3fR109 where a limited function is used.

**Expected behavior**
1. The path separator ';' should be used also when compiling vsg with MinGW.
2. The mentioned functions should be also used when compiling with MinGW

**Desktop (please complete the following information):**
 - OS: openSUSE
 - Version: Leap 15.4
 - MinGW header/libraries version 10
 - MinGW C++ Compiler 12.2

I've just tried the operator trick:

$ git diff
diff --git a/include/vsg/io/Path.h b/include/vsg/io/Path.h
index dcc6e0ac..5614aadb 100644
--- a/include/vsg/io/Path.h
+++ b/include/vsg/io/Path.h
@@ -121,6 +121,7 @@ namespace vsg
         inline const string_type& native() const noexcept { return _string; }
         inline operator const string_type&() const noexcept { return _string; }
         inline const value_type* c_str() const noexcept { return _string.c_str(); }
+        inline operator const value_type* () const noexcept { return _string.c_str(); }
 
         reference operator[](size_type pos) { return _string[pos]; }
         const_reference operator[](size_type pos) const { return _string[pos]; }

But I then get a compile error:

[ 56%] Building CXX object src/vsg/CMakeFiles/vsg.dir/text/Font.cpp.o
/home/robert/Dev/VulkanSceneGraph/src/vsg/io/VSG.cpp: In member function ‘virtual vsg::ref_ptr<vsg::Object> vsg::VSG::read(const vsg::Path&, vsg::ref_ptr<const vsg::Options>) const’:
/home/robert/Dev/VulkanSceneGraph/src/vsg/io/VSG.cpp:102:73: error: call of overloaded ‘basic_ifstream(vsg::Path&, std::_Ios_Openmode)’ is ambiguous
  102 |         std::ifstream fin(filenameToUse, std::ios::in | std::ios::binary);
      |                                                                         ^
In file included from /home/robert/Dev/VulkanSceneGraph/include/vsg/io/AsciiInput.h:21,
                 from /home/robert/Dev/VulkanSceneGraph/src/vsg/io/VSG.cpp:14:
/usr/include/c++/11/fstream:567:7: note: candidate: ‘std::basic_ifstream<_CharT, _Traits>::basic_ifstream(const string&, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::string = std::__cxx11::basic_string<char>; std::ios_base::openmode = std::ios_base::openmode]’
  567 |       basic_ifstream(const std::string& __s,
      |       ^~~~~~~~~~~~~~
/usr/include/c++/11/fstream:534:7: note: candidate: ‘std::basic_ifstream<_CharT, _Traits>::basic_ifstream(const char*, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::ios_base::openmode = std::ios_base::openmode]’
  534 |       basic_ifstream(const char* __s, ios_base::openmode __mode = ios_base::in)
      |       ^~~~~~~~~~~~~~
/home/robert/Dev/VulkanSceneGraph/src/vsg/io/VSG.cpp: In member function ‘virtual bool vsg::VSG::write(const vsg::Object*, const vsg::Path&, vsg::ref_ptr<const vsg::Options>) const’:
/home/robert/Dev/VulkanSceneGraph/src/vsg/io/VSG.cpp:183:70: error: call of overloaded ‘basic_ofstream(const vsg::Path&, std::_Ios_Openmode)’ is ambiguous
  183 |         std::ofstream fout(filename, std::ios::out | std::ios::binary);

To resolve this I have to comment out the 

       //inline operator const string_type&() const noexcept { return _string; }

But this then breaks the compile of tile.cpp:

/home/robert/Dev/VulkanSceneGraph/src/vsg/io/tile.cpp: In member function ‘virtual vsg::ref_ptr<vsg::Object> vsg::tile::read(const vsg::Path&, vsg::ref_ptr<const vsg::Options>) const’:
/home/robert/Dev/VulkanSceneGraph/src/vsg/io/tile.cpp:113:70: error: no matching function for call to ‘std::__cxx11::basic_stringstream<char>::basic_stringstream(vsg::Path&)’
  113 |         std::basic_stringstream<vsg::Path::value_type> sstr(tile_info);
      |                                                                      ^

This can be resolved by using tile_info.native()...

This could end up a bit convoluted so I'll create a branch to experiment with first and see if I can get it building with OK then let you review the branch.

--- a/include/vsg/io/Path.h
+++ b/include/vsg/io/Path.h
@@ -31,7 +31,7 @@ namespace vsg
     class VSG_DECLSPEC Path
     {
     public:
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
         using value_type = wchar_t;
         static constexpr value_type windows_separator = L'\\';
         static constexpr value_type posix_separator = L'/';
@@ -121,6 +121,9 @@ namespace vsg
         inline const string_type& native() const noexcept { return _string; }
         inline operator const string_type&() const noexcept { return _string; }
         inline const value_type* c_str() const noexcept { return _string.c_str(); }
+#if defined(__MINGW32__)
+        inline operator const value_type* () const noexcept { return _string.c_str(); }
+#endif
 
         reference operator[](size_type pos) { return _string[pos]; }
         const_reference operator[](size_type pos) const { return _string[pos]; }

--- a/src/vsg/io/FileSystem.cpp
+++ b/src/vsg/io/FileSystem.cpp
@@ -51,7 +51,7 @@ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLI
 
 using namespace vsg;
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
 const char envPathDelimiter = ';';
 #else
 const char envPathDelimiter = ':';

--- a/src/vsg/io/FileSystem.cpp
+++ b/src/vsg/io/FileSystem.cpp
@@ -59,7 +59,7 @@ const char envPathDelimiter = ':';
 
 std::string vsg::getEnv(const char* env_var)
 {
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
     char env_value[4096];
     std::size_t len;
     if (auto error = getenv_s(&len, env_value, sizeof(env_value) - 1, env_var); error != 0 || len == 0)
@@ -77,7 +77,7 @@ Paths vsg::getEnvPaths(const char* env_var)
 {
     if (!env_var) return {};
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
     char env_value[4096];
     std::size_t len;
     if (auto error = getenv_s(&len, env_value, sizeof(env_value) - 1, env_var); error != 0 || len == 0)
@@ -116,7 +116,7 @@ Paths vsg::getEnvPaths(const char* env_var)
 
 FileType vsg::fileType(const Path& path)
 {
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
     struct __stat64 stbuf;
     if (_wstat64(path.c_str(), &stbuf) != 0) return FILE_NOT_FOUND;
 #elif defined(__APPLE__)
@@ -137,7 +137,7 @@ FileType vsg::fileType(const Path& path)
 
 bool vsg::fileExists(const Path& path)
 {
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
     return _waccess(path.c_str(), 0) == 0;
 #else
     return access(path.c_str(), F_OK) == 0;
@@ -292,10 +292,8 @@ bool vsg::makeDirectory(const Path& path)
             continue;
         }
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
         if (int status = _wmkdir(directory_to_create.c_str()); status != 0)
-#elif defined(__MINGW32__)
-        if (int status = mkdir(directory_to_create.c_str()); status != 0)
 #else // POSIX
         if (int status = mkdir(directory_to_create.c_str(), 0755); status != 0)
 #endif
@@ -366,7 +364,7 @@ Path vsg::executableFilePath()
 
 FILE* vsg::fopen(const Path& path, const char* mode)
 {
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
     std::wstring wMode;
     convert_utf(mode, wMode);
 
@@ -381,7 +379,7 @@ FILE* vsg::fopen(const Path& path, const char* mode)
 #endif
 }
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || defined(__MINGW32__)
 // Microsoft API for reading directories
 Paths vsg::getDirectoryContents(const Path& directoryName)
 {

